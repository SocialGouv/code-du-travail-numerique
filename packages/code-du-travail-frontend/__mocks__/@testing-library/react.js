/* eslint-disable import/export */
import { theme } from "@socialgouv/cdtn-ui";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
