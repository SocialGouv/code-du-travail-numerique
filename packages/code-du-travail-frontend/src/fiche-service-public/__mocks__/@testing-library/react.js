import React from "react";
import PropTypes from "prop-types";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "@socialgouv/cdtn-ui";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={theme.colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
