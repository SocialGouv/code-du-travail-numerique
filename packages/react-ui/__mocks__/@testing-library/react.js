import { render } from "@testing-library/react";
import PropTypes from "prop-types";
import React from "react";
import { ThemeProvider } from "styled-components";

import { colors } from "../../src/theme";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={colors}>
      <>{children}</>
    </ThemeProvider>
  );
};

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
