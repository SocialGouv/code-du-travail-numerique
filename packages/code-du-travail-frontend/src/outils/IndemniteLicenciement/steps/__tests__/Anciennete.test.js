import { fireEvent, render } from "@testing-library/react";
import arrayMutators from "final-form-arrays";
import React from "react";
import { Form } from "react-final-form";

import {
  computeAnciennete,
  computeSalaires,
  StepAnciennete,
} from "../Anciennete";

function renderForm(data) {
  return render(
    <Form
      mutators={{
        ...arrayMutators,
      }}
      validate={StepAnciennete.validate}
      initialValues={{ ...data }}
      onSubmit={jest.fn()}
    >
      {({ form, handleSubmit }) => (
        <form onSubmit={handleSubmit}>
          <StepAnciennete form={form} />
          <button data-testid="nextBt">suivant</button>
        </form>
      )}
    </Form>
  );
}

describe("<Anciennete />", () => {
  it("should render", () => {
    const { container } = renderForm({});
    expect(container).toMatchSnapshot();
  });

  it("should display error if dateEntre after dateSortie", async () => {
    const { getByLabelText, container } = renderForm({
      dateEntree: "2018-04-02",
    });
    const dateSortie = getByLabelText(/date de sortie/i);
    fireEvent.change(dateSortie, { target: { value: "2018-03-01" } });
    expect(
      container
        .querySelector("input[name=dateSortie]")
        .parentElement.nextSibling.textContent.trim()
    ).toMatch("La date de sortie doit se situer après le 02 April 2018");
  });

  it("should display error if dateEntree after dateNotif", () => {
    const { container, getByLabelText } = renderForm({
      dateEntree: "2018-05-01",
    });
    const dateNotif = getByLabelText(/licenciement/i);
    fireEvent.change(dateNotif, { target: { value: "2018-04-02" } });
    expect(
      container
        .querySelector("input[name=dateNotification]")
        .parentElement.nextSibling.textContent.trim()
    ).toMatch("La date de notification doit se situer après la date d’entrée");
  });

  it("should display error if dateNotif after dateSortie", () => {
    const { container, getByLabelText } = renderForm({
      dateSortie: "2018-04-01",
    });
    const dateNotif = getByLabelText(/licenciement/i);
    fireEvent.change(dateNotif, { target: { value: "2018-05-02" } });
    expect(
      container
        .querySelector("input[name=dateNotification]")
        .parentElement.nextSibling.textContent.trim()
    ).toMatch("La date de notification doit se situer avant la date de sortie");
  });

  it("should display error if dateNotif is older than 18 months", () => {
    const { container, getByLabelText } = renderForm({
      dateSortie: "2022-04-01",
    });
    const dateNotif = getByLabelText(/licenciement/i);
    fireEvent.change(dateNotif, { target: { value: "2018-02-02" } });
    expect(
      container
        .querySelector("input[name=dateNotification]")
        .parentElement.nextSibling.textContent.trim()
    ).toMatch(
      "La date de notification doit se situer dans les 18 derniers mois"
    );
  });

  it("should display error if anciennté < 8mois", () => {
    const { getByText, getByLabelText } = renderForm({
      dateEntree: "2018-04-02",
      dateNotification: "2018-09-02",
    });
    const dateSortie = getByLabelText(/date de sortie/i);
    fireEvent.change(dateSortie, { target: { value: "2018-12-31" } });
    expect(getByText(/8 mois d’ancienneté/).textContent.trim()).toMatch(
      "L’indemnité de licenciement est dûe au-delà de 8 mois d’ancienneté"
    );
  });
});

describe("computeAncienneté", () => {
  it("should compute ancienneté", () => {
    expect(
      computeAnciennete({ dateEntree: "2017-04-01", dateSortie: "2018-04-01" })
    ).toEqual(1);
  });
  it("should compute ancienneté with periods of absence", () => {
    expect(
      computeAnciennete({
        absencePeriods: [{ duration: "6", type: "Congés sans solde" }],
        dateEntree: "2017-04-01",
        dateSortie: "2018-04-01",
      })
    ).toEqual(0.5);
  });
  it("should compute ancienneté with periods of absence divided by two for Congé parental", () => {
    expect(
      computeAnciennete({
        absencePeriods: [{ duration: "6", type: "Congé parental d'éducation" }],
        dateEntree: "2016-04-01",
        dateSortie: "2018-04-01",
      })
    ).toEqual(1.75);
  });
});

describe("computeSalaires", () => {
  it("should compute salaires periods only if there is previous salaires", () => {
    expect(
      computeSalaires({
        dateEntree: "2016-04-01",
        dateNotification: "2018-04-01",
        dateSortie: "2018-05-01",
        hasSameSalaire: false,
        salaires: [],
      }).length
    ).toEqual(12);
  });
  it("should restore previous salaires periods", () => {
    expect(
      computeSalaires({
        dateEntree: "2018-01-01",
        dateNotification: "2018-09-01",
        dateSortie: "2018-10-01",
        hasSameSalaire: false,
        salaires: [
          { label: "août 2018", salary: 2000 },
          { label: "juillet 2018", salary: 1450 },
          { label: "juin 2018", salary: 2000 },
          { label: "mai 2018", salary: 1500 },
          { label: "avril 2018", salary: 2000 },
          { label: "mars 2018", salary: 1400 },
          { label: "février 2018", salary: 2000 },
          { label: "janvier 2018", salary: 1500 },
        ],
      })
    ).toEqual([
      { label: "août 2018", salary: 2000 },
      { label: "juillet 2018", salary: 1450 },
      { label: "juin 2018", salary: 2000 },
      { label: "mai 2018", salary: 1500 },
      { label: "avril 2018", salary: 2000 },
      { label: "mars 2018", salary: 1400 },
      { label: "février 2018", salary: 2000 },
      { label: "janvier 2018", salary: 1500 },
    ]);
  });
  it("should not compute salaires periods if hasSameSalaire is true", () => {
    expect(
      computeSalaires({
        dateEntree: "2016-04-01",
        dateNotification: "2018-04-01",
        dateSortie: "2018-05-01",
        hasSameSalaire: true,
      })
    ).toBe(null);
  });
});
