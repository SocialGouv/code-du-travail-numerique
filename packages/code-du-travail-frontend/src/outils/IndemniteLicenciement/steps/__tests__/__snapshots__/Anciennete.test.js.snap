// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anciennete /> should render 1`] = `
.c2 {
  color: #3e486e;
  font-size: 1.6rem;
  font-weight: 400;
}

.c4 {
  position: relative;
  display: inline-block;
}

.c6 {
  width: 100%;
  height: 5.4rem;
  padding: 0 2rem;
  padding-right: 2rem;
  color: #3e486e;
  font-weight: normal;
  font-size: 1.6rem;
  font-family: "Open Sans",sans-serif;
  font-style: normal;
  line-height: inherit;
  text-align: left;
  background: #fff;
  border: 1px solid transparent;
  border-color: transparent;
  border-radius: 0.6rem;
  box-shadow: 0 1rem 2rem rgba(121,148,212,0.2);
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c6::-webkit-outer-spin-button,
.c6::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c6::-webkit-calendar-picker-indicator {
  display: block;
  width: 3.2rem;
  height: 3.2rem;
  margin-right: -2rem;
  color: rgba(0,0,0,0);
  background-color: #9298af;
  cursor: pointer;
  opacity: 1;
  -webkit-mask-image: url("data:image/svg+xml;,%3Csvg%20fill%3D%22none%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2032%2032%22%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M20.947%207.601h1.601c1.472%200%202.521%201.198%202.521%202.669v2.306l-.001%2010.392c0%201.472-1.049%202.669-2.521%202.669H8.669A2.672%202.672%200%20016%2022.967l.001-12.697a2.672%202.672%200%20012.67-2.669h1.331V6h.999v1.601H20V6h.947v1.601zm1.6%2017.036c.883%200%201.454-.786%201.454-1.67v-10.33h-17L7%2022.967c0%20.884.785%201.67%201.668%201.67h13.879zm-15.548-13h17.002l.001-1.367c0-.883-.57-1.601-1.453-1.601h-1.6v1.068H20V8.669h-9v1.068h-1V8.669H8.669C7.786%208.669%207%209.387%207%2010.27v1.367zm9.677%206.277h-2.135a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203h-2.135v2.135h2.136v-2.135zm3.203%203.203h2.135a1.07%201.07%200%20001.068-1.068v-2.135a1.07%201.07%200%2000-1.068-1.068H19.88a1.07%201.07%200%2000-1.068%201.068v2.135a1.07%201.07%200%20001.068%201.068zm0-3.203h2.135l.001%202.135H19.88v-2.135zm-8.54%208.541H9.204a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203H9.204v2.135h2.136l-.001-2.135zm3.202%203.203h2.135a1.07%201.07%200%20001.068-1.068v-2.135a1.07%201.07%200%2000-1.068-1.068h-2.135a1.07%201.07%200%2000-1.068%201.068v2.135a1.07%201.07%200%20001.068%201.068zm0-3.203h2.135l.001%202.135h-2.136v-2.135zm7.473%203.203H19.88a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203H19.88v2.135h2.136v-2.135z%22%20fill%3D%22currentColor%22%2F%3E%3C%2Fsvg%3E");
  mask-image: url("data:image/svg+xml;,%3Csvg%20fill%3D%22none%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2032%2032%22%3E%3Cpath%20fill-rule%3D%22evenodd%22%20clip-rule%3D%22evenodd%22%20d%3D%22M20.947%207.601h1.601c1.472%200%202.521%201.198%202.521%202.669v2.306l-.001%2010.392c0%201.472-1.049%202.669-2.521%202.669H8.669A2.672%202.672%200%20016%2022.967l.001-12.697a2.672%202.672%200%20012.67-2.669h1.331V6h.999v1.601H20V6h.947v1.601zm1.6%2017.036c.883%200%201.454-.786%201.454-1.67v-10.33h-17L7%2022.967c0%20.884.785%201.67%201.668%201.67h13.879zm-15.548-13h17.002l.001-1.367c0-.883-.57-1.601-1.453-1.601h-1.6v1.068H20V8.669h-9v1.068h-1V8.669H8.669C7.786%208.669%207%209.387%207%2010.27v1.367zm9.677%206.277h-2.135a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203h-2.135v2.135h2.136v-2.135zm3.203%203.203h2.135a1.07%201.07%200%20001.068-1.068v-2.135a1.07%201.07%200%2000-1.068-1.068H19.88a1.07%201.07%200%2000-1.068%201.068v2.135a1.07%201.07%200%20001.068%201.068zm0-3.203h2.135l.001%202.135H19.88v-2.135zm-8.54%208.541H9.204a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203H9.204v2.135h2.136l-.001-2.135zm3.202%203.203h2.135a1.07%201.07%200%20001.068-1.068v-2.135a1.07%201.07%200%2000-1.068-1.068h-2.135a1.07%201.07%200%2000-1.068%201.068v2.135a1.07%201.07%200%20001.068%201.068zm0-3.203h2.135l.001%202.135h-2.136v-2.135zm7.473%203.203H19.88a1.07%201.07%200%2001-1.068-1.068v-2.135a1.07%201.07%200%20011.068-1.068h2.135a1.07%201.07%200%20011.068%201.068v2.135a1.07%201.07%200%2001-1.068%201.068zm0-3.203H19.88v2.135h2.136v-2.135z%22%20fill%3D%22currentColor%22%2F%3E%3C%2Fsvg%3E");
}

.c6:invalid {
  border-color: #eb5757;
}

.c6::-webkit-input-placeholder {
  color: #9298af;
}

.c6::-moz-placeholder {
  color: #9298af;
}

.c6:-ms-input-placeholder {
  color: #9298af;
}

.c6::placeholder {
  color: #9298af;
}

.c6:focus {
  border-color: #7994d4;
}

.c6:focus::-webkit-input-placeholder {
  color: transparent;
}

.c6:focus::-moz-placeholder {
  color: transparent;
}

.c6:focus:-ms-input-placeholder {
  color: transparent;
}

.c6:focus::placeholder {
  color: transparent;
}

.c5 {
  width: 20rem;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.6rem;
  cursor: pointer;
}

.c9 {
  position: relative;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: calc(1.6rem * 1.25);
  height: calc(1.6rem * 1.25);
  margin: calc(1.6rem / 5) calc(2 * 1.6rem / 3) 0 0;
  padding: 0;
  line-height: inherit;
  background: #fff;
  border: 1px solid #bbcadf;
  border-radius: 50%;
  box-shadow: none;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c9:before {
  position: absolute;
  top: 50%;
  left: 50%;
  display: none;
  width: calc(1.6rem / 1.5);
  height: calc(1.6rem / 1.5);
  background-color: #f66663;
  border-radius: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  content: "";
}

.c9:checked {
  background-color: #fff;
  border-color: #f66663;
}

.c9:checked:before {
  display: block;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-bottom: 0.8rem;
}

.c0 {
  margin-top: 3.2rem;
  margin-bottom: 1rem;
  color: #4d73b8;
  font-weight: 600;
  font-size: 1.8rem;
  font-family: "Open Sans",sans-serif;
}

.c1 {
  display: block;
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1.8rem;
  font-weight: 600;
  cursor: pointer;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 2rem;
}

@media (max-width:600px) {
  .c4 {
    width: 100%;
  }
}

@media (max-width:600px) {
  .c6 {
    padding: 0 1rem;
    padding-right: 2rem;
  }
}

@media (max-width:600px) {
  .c5 {
    width: 100%;
  }
}

@media (max-width:600px) {
  .c1 {
    font-size: 1.6rem;
  }
}

<div>
  <form>
    <h2
      class="c0"
    >
      Dates d’entrée et de sortie de l’entreprise
    </h2>
    <label
      class="c1"
      for="1"
    >
      Quelle est la date d’entrée dans l’entreprise ?
      <span
        class="c2"
      >
         (obligatoire)
      </span>
    </label>
    <div
      class="c3"
    >
      <span
        class="c4 c5"
      >
        <input
          class="c6"
          id="1"
          max="2099-12-31"
          min="1900-01-01"
          name="dateEntree"
          placeholder=" jj/mm/yyyy"
          type="date"
          value=""
        />
      </span>
    </div>
    <label
      class="c1"
      for="2"
    >
      Quelle est la date de notification du licenciement ?
      <span
        class="c2"
      >
         (obligatoire)
      </span>
    </label>
    <div
      class="c3"
    >
      <span
        class="c4 c5"
      >
        <input
          class="c6"
          id="2"
          max="2099-12-31"
          min="1900-01-01"
          name="dateNotification"
          placeholder=" jj/mm/yyyy"
          type="date"
          value=""
        />
      </span>
    </div>
    <label
      class="c1"
      for="3"
    >
      Quelle est la date de sortie de l’entreprise ?
      <span
        class="c2"
      >
         (obligatoire)
      </span>
    </label>
    <div
      class="c3"
    >
      <span
        class="c4 c5"
      >
        <input
          class="c6"
          id="3"
          max="2099-12-31"
          min="1900-01-01"
          name="dateSortie"
          placeholder=" jj/mm/yyyy"
          type="date"
          value=""
        />
      </span>
    </div>
    <h2
      class="c0"
    >
      Période d’absence prolongée
    </h2>
    <label
      class="c1"
    >
      Y a-t-il eu des absences de plus d’un mois durant le contrat de travail ?
      <span
        class="c2"
      >
         (obligatoire)
      </span>
    </label>
    <div
      class="c7"
    >
      <label
        class="c8"
        for="hasAbsenceProlonge-oui"
      >
        <input
          class="c9"
          id="hasAbsenceProlonge-oui"
          name="hasAbsenceProlonge"
          type="radio"
          value="true"
        />
        Oui
      </label>
      <label
        class="c8"
        for="hasAbsenceProlonge-non"
      >
        <input
          class="c9"
          id="hasAbsenceProlonge-non"
          name="hasAbsenceProlonge"
          type="radio"
          value="false"
        />
        Non
      </label>
    </div>
    <button
      data-testid="nextBt"
    >
      suivant
    </button>
  </form>
</div>
`;
